### PR TITLE
[WIP] Adding Generic abstract class for tensors

### PR DESCRIPTION
**Pull request by interesaaat**
Trying to solve #18.

This will also help with PR #13 to avoid `object` type for inner tensors. 

This PR will go after PR #14 is merged.

### CODE REVIEW

**Comment by migueldeicaza** [comment]
I believe that this needs some discussion, which is why I had opened issue #18 as "Ponder".

I am not sure that `Tensor<T>` is the way to go, I mentioned that on #18, the reason being that Tensor<T> is not that different from DoubleTensor, in that it is difficult to change the code once a decision has made - the change has to take place everywhere.

The convenience of the Python API is that the tensor is typed at creation time, and checked when the operations are used, not at compile time.   So it is simple to change the models by merely changing the creation of the tensor, rather than changing every occurrence of the type (now, Pytorch also supports a global variable to change the default, but I think that is sort of gross).

One alternative would be to have a `Tensor` type, plus an additional `Tensor<T>`, to let folks pick their poison.

**Comment by interesaaat** [comment]
> I believe that this needs some discussion, which is why I had opened issue #18 as "Ponder".
> 
> I am not sure that `Tensor<T>` is the way to go, I mentioned that on #18, the reason being that Tensor is not that different from DoubleTensor, in that it is difficult to change the code once a decision has made - the change has to take place everywhere.
> 
> The convenience of the Python API is that the tensor is typed at creation time, and checked when the operations are used, not at compile time. So it is simple to change the models by merely changing the creation of the tensor, rather than changing every occurrence of the type (now, Pytorch also supports a global variable to change the default, but I think that is sort of gross).
> 
> One alternative would be to have a `Tensor` type, plus an additional `Tensor<T>`, to let folks pick their poison.

Ok now I got your "even better would be to have the `Tensor` not surface a `Tensor<T>`". I would rather say that this is related to the fact that Python is dynamically typed. Yes it is definitely convenient being able to easily swap types, but my experience (or, I would say, what I learned from ML.NET folks) is that it is more practical having strictly typed code with errors thrown at compile time, rather then runtime errors (eventually thrown after minutes, especially for training).

**Comment by migueldeicaza** [comment]
More often than not, the problem in these situations is mismatches on shape of the tensor, and we have no way of expressing this class of problems.   I agree with you that the type system would help, but in this case, it is catching a minor problem, while leaving the real problem unsolved.

**Comment by interesaaat** [comment]
I think we might be able to address the shape problem using value tuples in c#. We can definitely specify labels for dimensions in this way, but probably also lengths and related operation checks.

**Comment by migueldeicaza** [comment]
I do not understand what you have in mind, can you share?

**Comment by interesaaat** [comment]
> I do not understand what you have in mind, can you share?

I was thinking about something like adding a second constraint type to the tensor class in order to force static type checking of operators. Something like `Tensor<T, U> where U : Tuple<Dimension>`. I tried to implement something but the number of classes blows up because we need to enumerate all possible number of dimensions (unless there is a way to just say that `U` is a tuple, without explicitly saying with how many elements. You can do this in [scala](https://github.com/ctongfei/nexus) for instance.)

**Comment by NiklasGustafsson** [comment]
In order to capture shape as part of the type using tuples, you'd need something like C++'s variadic constructs, which isn't likely to happen in C#. Even Julia, which I consider almost ideal for numerics, doesn't try to make the actual shape part of the type, just the number of dimensions.

**Comment by dpthayer** [comment]
This problem is effectively resolved by using a Dependently Typed Language of which one of the premier examples is C++.  If you build the Tensor Library as a C++ Class you could pass template parameters to it such as the Number of dimensions and the size of each dimension and have lightning fast tensor operations at the machine level.

**Comment by dpthayer** [comment]
Something like the following
Tensor CreateTensor(uint dtype, uint rank, uint shape)
{
      //call to low level library
      return new NumericTenser<dtype, rank,shape>() as Tenser;
}

**Comment by dsyme** [comment]
@interesaaat I think we can close this?  Adding the generic baseclass is not particularly easy.

It could be done, but most of the operations would be TorchTensor -> TorchTensor with no generics statically propagated.

**Comment by dsyme** [comment]
p.s. I notice there was a use of SafeHandle which may be of interest here to put back into the code separately

**Comment by interesaaat** [comment]
Yes let's close this one. Even because now `TorchTensor` is a `struct` for performance reasons.
